PR: https://github.com/odoo/odoo/pull/155918

From: 1389a7298496a718937c2f9c50aa8f813640b9bd
From: dane@odoo.com
Date: 2024-04-24 16:29:00

Structural Changes: 4
Total Changes: 25

[IMP] base: contact tags improvements

The minor improvements are added to res.partner.category model:
1) grouping became possible based on color.
2) list view became multi editable.
3) quick search supports searching based on display_name
4) allow users to edit and create tags from res.partner view.

task-3775355

closes odoo/odoo#155918

Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/odoo/addons/base/models/res_partner.py
+++ b/odoo/addons/base/models/res_partner.py
@@ -112,9 +112,10 @@ class PartnerCategory(models.Model):
     def _get_default_color(self):
         return randint(1, 11)
 
-    name = fields.Char(string='Tag Name', required=True, translate=True)
-    color = fields.Integer(string='Color', default=_get_default_color)
-    parent_id = fields.Many2one('res.partner.category', string='Parent Category', index=True, ondelete='cascade')
+    name = fields.Char('Name', required=True, translate=True)
+    display_name = fields.Char('Display Name', compute='_compute_display_name', search='_search_display_name')
+    color = fields.Integer(string='Color', default=_get_default_color, aggregator=False)
+    parent_id = fields.Many2one('res.partner.category', string='Category', index=True, ondelete='cascade')
     child_ids = fields.One2many('res.partner.category', 'parent_id', string='Child Tags')
     active = fields.Boolean(default=True, help="The active field allows you to hide the category without removing it.")
     parent_path = fields.Char(index=True)
@@ -147,6 +148,9 @@ class PartnerCategory(models.Model):
             domain = [('name', operator, name)] + domain
         return self._search(domain, limit=limit, order=order)
 
+    @api.model
+    def _search_display_name(self, operator, value):
+        return [('id', 'child_of', self._search([('name', operator, value)]))]
 
 class PartnerTitle(models.Model):
     _name = 'res.partner.title'

--- a/odoo/addons/base/views/res_partner_views.xml
+++ b/odoo/addons/base/views/res_partner_views.xml
@@ -237,7 +237,7 @@
                                 invisible="is_company"/>
                             <field name="active_lang_count" invisible="1"/>
                             <field name="lang" invisible="active_lang_count &lt;= 1"/>
-                            <field name="category_id" widget="many2many_tags" options="{'color_field': 'color', 'no_create_edit': True}"
+                            <field name="category_id" widget="many2many_tags" options="{'color_field': 'color'}"
                                    placeholder='e.g. "B2B", "VIP", "Consulting", ...'/>
                         </group>
                     </group>
@@ -651,8 +651,9 @@
             <field name="model">res.partner.category</field>
             <field eval="6" name="priority"/>
             <field name="arch" type="xml">
-                <tree string="Contact Tags">
-                    <field name="display_name"/>
+                <tree string="Contact Tags" editable="bottom" multi_edit="1" sample="1">
+                    <field name="name" placeholder='e.g. "Roadshow"'/>
+                    <field name="parent_id" string="Category" domain="[('id', '!=', id)]"/>
                     <field name="color" widget="color_picker"/>
                 </tree>
             </field>
@@ -663,10 +664,14 @@
             <field name="model">res.partner.category</field>
             <field name="arch" type="xml">
                 <search string="Search Partner Category">
-                    <field name="name"/>
-                    <field name="display_name"/>
+                    <field name="name" string="Name"/>
+                    <field name="display_name" string="Category" />
                     <separator/>
                     <filter string="Archived" name="inactive" domain="[('active', '=', False)]"/>
+                    <group string="Group By">
+                        <filter string="Category" name="group_parent_id" context="{'group_by': 'parent_id'}"/>
+                        <filter string="Color" name="group_color" context="{'group_by': 'color'}"/>
+                    </group>
                 </search>
             </field>
         </record>
